linux/mcompile/keymap/mc_import_rules.cpp

```diff
@@ -0,0 +1,651 @@
+/*
+  Name:             mc_import_rules
+  Copyright:        Copyright (C) SIL International.
+  Documentation:    
+  Description:      
+  Create Date:      3 Aug 2014
+
+  Modified Date:    6 Feb 2015
+  Authors:          mcdurdin
+  Related Files:    
+  Dependencies:     
+
+  Bugs:             
+  Todo:             
+  Notes:            
+  History:          03 Aug 2014 - mcdurdin - I4327 - V9.0 - Mnemonic layout compiler follow-up
+                    03 Aug 2014 - mcdurdin - I4353 - V9.0 - mnemonic layout recompiler mixes up deadkey rules
+                    31 Dec 2014 - mcdurdin - I4550 - V9.0 - logical flaw in mnemonic layout recompiler means that AltGr base keys are never processed
+                    06 Feb 2015 - mcdurdin - I4552 - V9.0 - Add mnemonic recompile option to ignore deadkeys
+*/
+
+#include <vector>
+#include <string>
+#include <stdio.h>
+//#include "km_types.h"
+#include "mc_kmxfile.h"
+#include "keymap.h"
+
+const int KMX_ShiftStateMap[] = {
+  ISVIRTUALKEY,
+  ISVIRTUALKEY | K_SHIFTFLAG,
+  ISVIRTUALKEY | K_CTRLFLAG,
+  ISVIRTUALKEY | K_SHIFTFLAG | K_CTRLFLAG,
+  0,
+  0,
+  ISVIRTUALKEY | RALTFLAG,
+  ISVIRTUALKEY | RALTFLAG | K_SHIFTFLAG,
+  0,
+  0
+};
+
+DeadKey::DeadKey(KMX_WCHAR deadCharacter) {
+  this->m_deadchar = deadCharacter;
+}
+
+KMX_WCHAR DeadKey::KMX_DeadCharacter() {
+  return this->m_deadchar;
+}
+
+void DeadKey::KMX_AddDeadKeyRow(KMX_WCHAR baseCharacter, KMX_WCHAR combinedCharacter) {
+  this->m_rgbasechar.push_back(baseCharacter);
+  this->m_rgcombchar.push_back(combinedCharacter);
+}
+
+bool DeadKey::KMX_ContainsBaseCharacter(KMX_WCHAR baseCharacter) {
+  std::vector<KMX_WCHAR>::iterator it;
+  for(it=this->m_rgbasechar.begin(); it<m_rgbasechar.end(); it++) {
+    if(*it == baseCharacter) {
+      return true;
+    }
+  }
+  return false;
+}
+
+int KMX_ToUnicodeEx(guint keycode, PKMX_WCHAR pwszBuff, int shift_state_pos, int caps,GdkKeymap *keymap) {
+  GdkKeymapKey *maps;
+  guint *keyvals;
+  gint count;
+
+  if (!gdk_keymap_get_entries_for_keycode(keymap, keycode, &maps, &keyvals, &count))
+    return 0;
+
+  if (!(shift_state_pos <= count))
+   return 0;
+
+  if (!(keycode <= keycode_max))
+    return 0;
+
+  KMX_DWORD KeyVal= (KMX_DWORD) KMX_get_KeyVal_From_KeyCode(keymap, keycode, ShiftState(shift_state_pos), caps);
+  std::u16string str = convert_DeadkeyValues_To_U16str(KeyVal);
+  pwszBuff[0]= * (PKMX_WCHAR)  str.c_str();
```

Comment:
- I use a buffer here because on windows we use a buffer at this place and I am hesitant to change the code too much in order to keep things as similar as possible on all platforms
- We will not have to worry about surrogate pairs since mcompile is only used for latin scripts. 
- On windows we use ToUnicodeEx to obtain a character and a returnvalue rc that specifies if we found a dk/character/no translation. Again for reasons of similarity I called the equivalent Linux-function  KMX_ToUnicodeEx (even though it is not exactly the same as on windows) because on Linux we too need a character and the return value rc.

Reply:
see PR #11816

Reply:
Yes, I understand. However, a name implies a certain functionality, and so if the name contains Unicode the reader expects it to properly deal with Unicode which includes surrogate pairs. Without careful inspection of where this gets called from and what's passed in it's impossible to know that surrogate pairs are out of scope. I see three possible options to solve this:
- rename the method
- assert that str.c_str()[0] is not a surrogate, i.e. is not in the range 0xD800-0xDFFF (with a comment that those are surrogates).
- or at least a comment in the method that we intentionally don't deal with surrogate pairs since we only deal with latin scripts